ai_diffusion/comfy_client.py

```diff
@@ -340,9 +342,34 @@ async def clear_queue(self):
             except asyncio.QueueEmpty:
                 break
 
-        await self._post("queue", {"clear": True})
+        remote_ids = [await job.get_remote_id() for job in self._jobs]
+        await self._post("api/queue", {"delete": remote_ids})
+
         self._jobs.clear()
 
+    async def _poll_server_queue(self):
+        queue = await self._get("api/queue")
+        server_jobs = queue["queue_running"] + queue["queue_pending"]
+        # why are they unsorted to start with...?
+        server_jobs = sorted(server_jobs, key=itemgetter(0))
+        server_jobs = [entry[1] for entry in server_jobs]
+        if not (self._jobs or self._active):
+            await self._report(ClientEvent.foreign_jobs, "", foreign_jobs=len(server_jobs))
+            return
+
+        if self._active:
+            first_remote_id = await self._active.get_remote_id()
+        else:
+            first_remote_id = await self._jobs[0].get_remote_id()
+        # if we got it from _jobs or _active, this field must have been set (in _run_job).
+        first_remote_id = util.ensure(first_remote_id)
+        try:
+            offset = server_jobs.index(first_remote_id)
+        except ValueError:
+            # probably just haven't gotten the notification yet
+            return
+        await self._report(ClientEvent.foreign_jobs, "", foreign_jobs=offset)
```

Comment:
Does it make sense to report the ID here? The UI would consider the job to be executing while waiting for its turn in the queue, which I think is how it already worked for the cloud client

Reply:
Yea the cancel is a bit ambiguous in this case. But from user perspective I think it makes sense to track the active job (the earliest submitted by the user and not finished yet), and consider all follow-up jobs enqueued. The active job may be in a "waiting" state while the server is busy.

"Cancel active" should probably remove the active job from the server queue if it isn't processing yet in that case.

Reply:
I'm gonna leave that out for now cause I think you underestimate how ugly it's going to be. Right now the job queue tracks the server-side job queue pretty closely, which is why we can just call `interrupt()` to clear out the running job. But that's state that's distributed over like three classes, ie. client, jobs and the comfy server. If we start pretending that a queued job is an active one, even in the job queue, then the clients have to keep their own account of what's going on and redirect interrupt() to *sometimes* do `/interrupt` and sometimes `/api/queue: delete`.

Reply:
Yes I thought the interrupt call might just become:
```
if self._active_job:
  self._post("/interrupt")
else:
  self._post("/api/queue", {"delete": self._jobs[0].remote_id})
```
(something like that)

But maybe not?

Reply:
Idk, if anything, shouldn't it be done in model, ie. `cancel`? I just want to minimize the amount of state that is lying.

Reply:
It's not lying, only a different perspective.

The Model reflects what the user is interested in. The UI presents it. The Client maps the inner workings of whatever server it's written for to the outside model for easy consumption. Knowledge about inner workings of the server shouldn't leave the Client.

UI absolutely wants to paper over details users aren't interested in. I disagree on "close to the surface", complexity should be dealt with at the lowest layer possible, rather than leaking to upper layers, so they don't have to bother with it.

As a user I hit Generate, and from that moment I've started a job to create an image. It may involve several stages (waiting for spot in queue, uploading, generating, downloading) and it's done when I can see the image.
* I can enqueue further jobs, but those won't start until the current job is done. 
* I can cancel the current one if I'm no longer interested in that particular image, and it should interrupt processing at whatever stage it is in (I don't care).
* I can cancel enqueued jobs, and it should remove them from the queue without affecting the current job.

The client is responsible for communication with the server, so it needs some more detailed representation of a job. It may not consider a job active unless it's actually executing, so it knows what to do when it's supposed to be canceled (dequeue vs interrupt).

I think that's the desired behavior at least. Maybe the naming in the code could reflect it better?

Reply:
I think I've figured out why this bewilders me so. In that case, isn't `job.any_executing` silly? We shouldn't care if a job is "executing" at all, and the cancel button should never be disabled unless there's *no* jobs in the queue. `any_executing` should be removed, and `JobState` should just be `queued, finished, cancelled`.

I think that's a valid refactor, and I'll do it if you think it's a blocker, but at any rate it does not belong in this PR.

Reply:
I think that, if the "cancel first" button is always shown, and always removes the first element, there genuinely isn't any code left that conditions on `JobState.executing`. Am I missing any?

Reply:
You mean any(queued) and any(executing) are equivalent? Mostly I guess. If you start a job and it errors early (connection failure or server enqueue failing directly) it never goes into executing.

Reply:
I mean, does it matter at that point? With that refactor, what is actually still dependent on executing?

Reply:
I'm sure making executing implicit (oldest non-finished-non-cancelled job) works too, but I kinda like that it relies on client feedback and is an explicit state. I don't see a strong argument one way or the other, but if there's no strong reason to change, keep what already works.

Reply:
I'm just hopelessly confused now. If you like that job state relies on client feedback, why do you want me to mark the first job, that the client has not started executing yet, as executing?

Reply:
Because I consider standing in line as part of executing my job! You know, pushing a little, trying to advance a spot or two when those in front are looking the other way. Things a good client does.

Call it "active" or "in_progress" or something if executing is so confusing. The important part is that the job is the focus of attention and something is being done about it. That's the job I care about, and which the UI progress bars and indicators refer to. Surely that's worth a special state? (compared to all the other jobs that are either old, finished and forgotten or distant future)